server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.datastreams.autosharding;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.DataStream;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.IndexMetadataStats;
+import org.elasticsearch.cluster.metadata.IndexWriteLoad;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.features.FeatureService;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.index.Index;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.OptionalDouble;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+
+import static org.elasticsearch.action.datastreams.autosharding.AutoShardingResult.NOT_APPLICABLE_RESULT;
+
+/**
+ * Calculates the optimal number of shards the data stream write index should have based on the indexing load.
+ */
+public class DataStreamAutoShardingService {
+
+    private static final Logger logger = LogManager.getLogger(DataStreamAutoShardingService.class);
+    public static final String DATA_STREAMS_AUTO_SHARDING_ENABLED = "data_streams.auto_sharding.enabled";
+
+    public static final NodeFeature DATA_STREAM_AUTO_SHARDING_FEATURE = new NodeFeature("data_stream.auto_sharding");
+
+    public static final Setting<List<String>> DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING = Setting.listSetting(
+        "data_streams.auto_sharding.excludes",
+        List.of("*"),
+        Function.identity(),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * Represents the minimum amount of time between two scaling events if the next event will increase the number of shards.
+     */
+    public static final Setting<TimeValue> DATA_STREAMS_AUTO_SHARDING_INCREASE_SHARDS_COOLDOWN = Setting.timeSetting(
+        "data_streams.auto_sharding.increase_shards.cooldown",
+        TimeValue.timeValueMinutes(15),
```

Comment:
As we discussed earlier with @henningandersen, we could potentially lower this to be the same as the periodic polling for the data stream lifecycle service, which makes sense (so that during master failure we wait for at least N (default 5) minutes before scaling up shards.

The other question I have is whether we should just go ahead and make it a factor (still configurable) of the poll interval, so we could do something like `1.0` or `.9` and make it 90% of 5 minutes (so 270 seconds instead of 300).

I don't think we have to figure the relative part out in this PR necessarily (but I'd like Henning to comment on whether 5m is fine, since that's what we discussed earlier).

Reply:
++ note that my understanding was that we'll be configuring it to slightly lower than the DSL poll interval - i.e. 4m - to make sure we **do** auto shard on subsequent data stream lifecycle runs (barring master failovers or shutdowns)

Reply:
I've pushed 270 seconds as part of this PR but happy to amend it as needed